### ISSUE TITLE
e2e test coverage for Grafana dashboards

### ISSUE THREAD

**Issue by heanlan** [antrea-io/theia#52]
**Background** 

Code provided by us/target code to be tested on:
All dashboard JSON files, including panels configuration, ClickHouse SQL queries.

Grafana data logic flow:
- Refresh a dashboard
- Grafana loops through all the queries and calls the appropriate data source. This step is done by the third party datasource plugin. The datasource plugin takes the query as the input, send the query to the external datasource API, and return the query result data as the output to Grafana
- Grafana panel plugin takes the query result as input, render and display the graphs on the panel

**Solution1 - [Grafana HTTP API](https://grafana.com/docs/grafana/latest/developers/http_api/)**

The Grafana backend exposes an HTTP API, which is the same API that is used by the frontend to do everything from saving dashboards, creating users, and updating data sources.

Eligibilities relevant our use case: 
- get a list of dashboards/get a single dashboard by uid
- get a list of datasources/get a single data source by uid

Limitation

However, in the Grafana HTTP API, it doesn't include an API that can execute a query and get the query result.
Ref issue1: https://community.grafana.com/t/backend-api-to-get-query-result/67293/2
Ref issue2: https://community.grafana.com/t/grafana-http-api-to-get-panel-json-data/63901/3
Ref issue3: https://community.grafana.com/t/dashboard-api-returns-query-results-as-well/5556/7

If we want to verify the query result, one alternative is: Send request to Grafana dashboard API, get the dashboard JSON file, extract the query from the dashboard JSON, and run the query independently against the datasource.

**Solution2 - [Grafana e2e package](https://github.com/grafana/grafana/blob/main/contribute/style-guides/e2e.md)**

It provides a package built on top of [Cypress](https://www.cypress.io/), which allow us to define some actions on the app and define the corresponding expected outcome. e.g. Open a dashboard, check correct panels are displayed. The package doesn't seem to have a clear documentation, but there are some example [test suites](https://github.com/grafana/grafana/tree/main/e2e) to get start with. 

Limitation

- When we run tests in Cypress, it launch a browser for us, which doesn't fit our Kind test environment.
- Although it allows us to execute a query. Still, it cannot automatically retrieve and verify the query result. We need to compare by our own the testing panel visualization with expected panel screenshot. [Example](https://github.com/grafana/clickhouse-datasource/tree/5ff43fc6d609912e3ae1163bf19699d43e03d1a8/cypress-e2e)

**Comment by heanlan** [comment]
cc @dreamtalen @yanjunz97 @yuntanghsu @salv-orlando for discussion and evaluation. Any feedback would be much appreciated.

**Comment by heanlan** [comment]
> Is it possible to run e2e tests for [grafana-clickhouse datasource](https://github.com/grafana/clickhouse-datasource) directly to cover this missing part in the above approach?

Are you referring to this e2e test https://github.com/grafana/clickhouse-datasource/tree/5ff43fc6d609912e3ae1163bf19699d43e03d1a8/cypress-e2e ? It is built with Cypress, needs to be run in a browser.

**Comment by dreamtalen** [comment]
> > Is it possible to run e2e tests for [grafana-clickhouse datasource](https://github.com/grafana/clickhouse-datasource) directly to cover this missing part in the above approach?
> 
> Are you referring to this e2e test https://github.com/grafana/clickhouse-datasource/tree/5ff43fc6d609912e3ae1163bf19699d43e03d1a8/cypress-e2e ? It is built with Cypress, needs to be run in a browser.

No, I mean is it possible to test these steps `The datasource plugin takes the query as the input, send the query to the external datasource API, and return the query result data as the output to Grafana` directly? Without rendering dashboards and comparing screenshots in a browser.

**Comment by yuntanghsu** [comment]
I think the only difference between (a) and (b) is the `$timeFilter()`, which can be deleted? And we can just use the parameters in (a) for the $timeInterval, 

One thing comes up to my mind is that after clickhouse cluster merged, not sure if the original query are still usable.

I'm sending query to clickhouse db to retrieve metrics and verify their values. But for most of the queries, I need to add "cluster" to query in order to all the data in shards. e.g. "cluster('{cluster}', INFORMATION_SCHEMA.COLUMNS)"
If you only use distributed engine table, I think it might not be an issue?

**Comment by yanjunz97** [comment]
I do not have much to add to the discussion. Regarding to choices between (a) and (b), (b) looks better to me, as it seems to include the datasource plugin into the whole e2e tests.  But as it depends on the plugin API, let's wait to see the reply from them.

Besides, I'm curious about the fact that `rawQuery` is not always aligned with the `sqlQuery`. In this case, do we still get the result as we expected based on the `sqlQuery` we write? Does that mean `rawQuery` is not something the datasource plugin used to interact with the ClickHouse server? IMO it might be better if we can use what the plugin uses to talk to the ClickHouse server if we have to choose option (a).

**Comment by yanjunz97** [comment]
> One thing comes up to my mind is that after clickhouse cluster merged, not sure if the original query are still usable.
> 
> I'm sending query to clickhouse db to retrieve metrics and verify their values. But for most of the queries, I need to add "cluster" to query in order to all the data in shards. e.g. "cluster('{cluster}', INFORMATION_SCHEMA.COLUMNS)" If you only use distributed engine table, I think it might not be an issue?

I think the queries regarding the table flows and other MVs should have no difference between distributed table and stand alone table? Does "most of the queries" refer to system tables for things like storage consumption? I think even if we introduce ClickHouse cluster, we still need to support both cases, i.e., everything should work for both stand alone ClickHouse and ClickHouse cluster.

**Comment by yuntanghsu** [comment]
> I think the queries regarding the table flows and other MVs should have no difference between distributed table and stand alone table? Does "most of the queries" refer to system tables for things like storage consumption? I think even if we introduce ClickHouse cluster, we still need to support both cases, i.e., everything should work for both stand alone ClickHouse and ClickHouse cluster.

Yes, only for system tables. Adding "cluster" can make sure we have same behavior for standalone ClickHouse and ClickHouse cluster.

**Comment by heanlan** [comment]
Thanks @yanjunz97 and @yuntanghsu . For your questions on the `rawQuery`, `rawSql`, I have an answer now. 

As Yun-Tang might not know, for the datasource plugin, we were previously using the vertamedia-clickhouse-grafana and then switch to grafana-clickhouse-datasource. `grafana-clickhouse-datasource` only introduced the `rawSql`, while 
`vertamedia-clickhouse-grafana` introduced `formattedQuery`, `query`,  and `rawQuery` . So in our dashboard JSONs, they now have all these four "Query", which makes it confusing. Sorry for not discovering this issue earlier. I'll delete those unused `formattedQuery`, `query`,  and `rawQuery`  from dashboard JSONs. Ref issue: https://github.com/grafana/clickhouse-datasource/issues/144 

So for `grafana-clickhouse-datasource`, the only input is `rawSql`. They do a relative complex parsing/translating on the `rawSql` and send to the datasource. 

Another (possibly good) update is, the plugin team suggested we can retrieve the query results from `api/ds/query `. I'll spend time to verify whether that's what we want, as it definitely contradicts with what being told by the Grafana team 🙃 
https://grafana.com/docs/grafana/latest/developers/http_api/data_source/#query-a-data-source
